bitcoin/psbt.c

```diff
@@ -780,38 +774,23 @@ struct wally_psbt *fromwire_wally_psbt(const tal_t *ctx,
 	return psbt;
 }
 
-/* This only works on a non-final psbt because we're ALL SEGWIT! */
 void psbt_txid(const tal_t *ctx,
-	       const struct wally_psbt *psbt, struct bitcoin_txid *txid,
+	       const struct wally_psbt *psbt,
+		   struct bitcoin_txid *txid,
 	       struct wally_tx **wtx)
 {
 	struct wally_tx *tx;
+	int wally_err;
+	assert(psbt->version == 2);
 
-	/* You can *almost* take txid of global tx.  But @niftynei thought
-	 * about this far more than me and pointed out that P2SH
-	 * inputs would not be represented, so here we go. */
+	/* We rely on wally extractor to fill out all txid-related fields including scriptSigs */
 	tal_wally_start();
-	wally_tx_clone_alloc(psbt->tx, 0, &tx);
-
-	for (size_t i = 0; i < tx->num_inputs; i++) {
-		if (psbt->inputs[i].final_scriptsig) {
-			wally_tx_set_input_script(tx, i,
-						  psbt->inputs[i].final_scriptsig,
-						  psbt->inputs[i].final_scriptsig_len);
-		} else if (psbt->inputs[i].redeem_script) {
-			u8 *script;
-
-			/* P2SH requires push of the redeemscript, from libwally src */
-			script = tal_arr(tmpctx, u8, 0);
-			script_push_bytes(&script,
-					  psbt->inputs[i].redeem_script,
-					  psbt->inputs[i].redeem_script_len);
-			wally_tx_set_input_script(tx, i, script, tal_bytelen(script));
-		}
-	}
-	tal_wally_end_onto(ctx, tx, struct wally_tx);
+	wally_err = wally_psbt_extract(psbt, WALLY_PSBT_EXTRACT_NON_FINAL, &tx);
+	assert(wally_err == WALLY_OK);
+	wally_err = wally_tx_get_txid(tx, txid->shad.sha.u.u8, sizeof(txid->shad.sha.u.u8));
+	assert(wally_err == WALLY_OK);
+	tal_wally_end(ctx);
```

Comment:
:heart: Gotta love this from libwally!

Reply:
any time I don't have to re-implement something an angel gets its wings